File: DevEnvironment/Problems/src/main/java/problems/core/MinimumWindowSubstring.java
package problems.core;

import base.core.Conditions;
import base.core.Indexes;
import base.interfaces.ICalculator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * The MinimumWindowSubstring class implements a minimum windows substring.
 */
public final class MinimumWindowSubstring implements ICalculator<String> {
    private final char[] src;
    private final Set<Character> characters;

    /**
     * The MinimumWindowSubstring constructor.
     */
    public MinimumWindowSubstring(String s, String t) {
        Conditions.validateNotNull(
            s,
            "The s string.");

        Conditions.validateNotNull(
            t,
            "The t string.");

        this.src = s.toCharArray();
        this.characters = MinimumWindowSubstring.createCharacterSet(t);
    }

    /**
     * Given a string S and a string T, find the minimum window in S,
     * which will contain all the characters in T in complexity O(n).
     */
    @Override
    public String calculate() {
        if (this.characters.isEmpty() || this.src.length < this.characters.size()) {
            return null;
        }

        Map<Character, Integer> characterCounter = new HashMap<>();

        int endIndex = this.coverSourceString(
            this.src,
            0,
            this.src.length - 1,
            characterCounter);

        if (endIndex == -1) {
            return null;
        }

        int startIndex = this.shrinkSubString(
            this.src,
            0,
            endIndex,
            characterCounter);

        int minStartIndex = startIndex;
        int minLength = Indexes.size(startIndex, endIndex);

        for (int currIndex = endIndex + 1; currIndex < this.src.length; ++currIndex) {
            char currValue = this.src[currIndex];

            if (!this.characters.contains(currValue)) {
                continue;
            }

            updateCharacterCounter(
                characterCounter,
                currValue,
                1);

            startIndex = this.shrinkSubString(
                this.src,
                startIndex,
                currIndex,
                characterCounter);

            int currLength = Indexes.size(startIndex, currIndex);

            if (currLength < minLength) {
                minStartIndex = startIndex;
                minLength = currLength;
            }
        }

        return new String(this.src, minStartIndex, minLength);
    }

    /**
     * Covers the source string.
     * Returns -1 if the string is not able to be covered.
     */
    private int coverSourceString(
        char[] src,
        int startIndex,
        int endIndex,
        Map<Character, Integer> characterCounter) {

        for (int i = startIndex; i <= endIndex; ++i) {
            char currValue = src[i];

            if (this.characters.contains(currValue)) {
                updateCharacterCounter(characterCounter, currValue, 1);

                if (characterCounter.size() == this.characters.size()) {
                    return i;
                }
            }
        }

        return -1;
    }

    /**
     * Shrinks a sub-string.
     */
    private int shrinkSubString(
        char[] src,
        int startIndex,
        int endIndex,
        Map<Character, Integer> characterCounter) {

        int currIndex = startIndex;

        while (currIndex <= endIndex) {
            char currValue = src[currIndex];

            if (this.characters.contains(currValue)) {
                int currValueCount = characterCounter.get(currValue);
                assert(currValueCount >= 1);

                if (currValueCount == 1) {
                    break;
                }

                characterCounter.put(currValue, currValueCount - 1);
            }

            ++currIndex;
        }

        return currIndex;
    }

    /**
     * Updates the character counter.
     */
    private static void updateCharacterCounter(
        Map<Character, Integer> characterCounter,
        char value,
        int delta) {

        int currCount = characterCounter.getOrDefault(value, 0);
        currCount += delta;
        characterCounter.put(value, currCount);
    }

    /**
     * Creates a character set.
     */
    private static Set<Character> createCharacterSet(String str) {
        Set<Character> characters = new HashSet<>();

        for (char character : str.toCharArray()) {
            characters.add(character);
        }

        return characters;
    }
}

File: DevEnvironmentCLion/Problems/core/MinimumWindowSubstring.cpp
#include "PreCompiled.h"

#include "MinimumWindowSubstring.h"
#include "Dimensions.h"

using namespace problems;

/**
 * The MinimumWindowSubstring constructor.
 */
MinimumWindowSubstring::MinimumWindowSubstring(const std::string& s, const std::string& t) :
    m_src(s),
    m_characters(MinimumWindowSubstring::createCharacterSet(t))
{
}

/**
 * Given a string S and a string T, find the minimum window in S,
 * which will contain all the characters in T in complexity O(n).
 */
bool MinimumWindowSubstring::findMinimumWindow(std::string& result)
{
    if (m_characters.empty() || m_src.size() < m_characters.size()) {
        return false;
    }

    const std::string::value_type* srcStr = m_src.c_str();

    std::map<std::string::value_type, std::size_t> characterCounter;

    std::size_t endIndex;
    bool status = coverSourceString(
        srcStr,
        0,
        m_src.size() - 1,
        endIndex,
        characterCounter);

    if (!status)
    {
        return false;
    }

    std::size_t startIndex = shrinkSubString(
        srcStr,
        0,
        endIndex,
        characterCounter);

    std::size_t minStartIndex = startIndex;
    std::size_t minEndIndex = endIndex;

    std::size_t currIndex = endIndex + 1;

    while (currIndex < m_src.size())
    {
        std::string::value_type currValue = srcStr[currIndex];
        std::size_t endIndex = currIndex;

        if (m_characters.find(currValue) != m_characters.end())
        {
            MinimumWindowSubstring::updateCharacterCounter(
                characterCounter,
                currValue,
                1);

            startIndex = shrinkSubString(
                srcStr,
                startIndex,
                endIndex,
                characterCounter);

            std::size_t currLength = base::Dimensions::length(startIndex, endIndex);
            std::size_t minLength = base::Dimensions::length(minStartIndex, minEndIndex);

            if (currLength < minLength)
            {
                minStartIndex = startIndex;
                minEndIndex = endIndex;
            }
        }

        ++currIndex;
    }

    std::size_t minLength = base::Dimensions::length(minStartIndex, minEndIndex);
    result = m_src.substr(minStartIndex, minLength);

    return true;
}

/**
 * Covers the source string.
 * Returns -1 if the string is not able to be covered.
 */
bool MinimumWindowSubstring::coverSourceString(
    const std::string::value_type* str,
    std::size_t startIndex,
    std::size_t endIndex,
    std::size_t& coveredStringEndIndex,
    CharacterCounterMap& characterCounter)
{
    std::size_t currIndex = startIndex;

    while (currIndex <= endIndex)
    {
        std::string::value_type currValue = str[currIndex];

        if (m_characters.find(currValue) != m_characters.end())
        {
            MinimumWindowSubstring::updateCharacterCounter(characterCounter, currValue, 1);

            if (characterCounter.size() == m_characters.size())
            {
                coveredStringEndIndex = currIndex;
                return true;
            }
        }

        ++currIndex;
    }

    return false;
}

/**
 * Shrinks a sub string.
 */
 std::size_t MinimumWindowSubstring::shrinkSubString(
     const std::string::value_type* str,
     std::size_t startIndex,
     std::size_t endIndex,
     CharacterCounterMap& characterCounter)
 {
     std::size_t currIndex = startIndex;

     while (currIndex <= endIndex)
     {
         std::string::value_type currValue = str[currIndex];

         if (m_characters.find(currValue) != m_characters.end())
         {
             if (characterCounter[currValue] == 1)
             {
                 return currIndex;
             }
             else
             {
                MinimumWindowSubstring::updateCharacterCounter(characterCounter, currValue, -1);
             }
         }

         ++currIndex;
     }

     return currIndex;
 }

/**
 * Updates the character counter.
 */
void MinimumWindowSubstring::updateCharacterCounter(
    CharacterCounterMap& characterCounter,
    std::string::value_type character,
    int delta)
{
    std::size_t currCount = 0;
    CharacterCounterMap::const_iterator i = characterCounter.find(character);

    if (i != characterCounter.end()) {
        currCount = i->second;
    }

    currCount += delta;

    characterCounter[character] = currCount;
}

/**
 * Creates a character set.
 */
std::set<std::string::value_type> MinimumWindowSubstring::createCharacterSet(const std::string& str) {
    std::set<std::string::value_type> characterSet;

    for (std::string::value_type character : str) {
        characterSet.insert(character);
    }

    return characterSet;
}
